services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/helpers/Wrappers.groovy

```diff
@@ -0,0 +1,93 @@
+package org.openkilda.functionaltests.helpers
+
+import groovy.util.logging.Slf4j
+import org.junit.runners.model.MultipleFailureException
+
+@Slf4j
+class Wrappers {
+
+    /**
+     * Retries some operation for a given number of times. Retry is considered failed if any throwable is thrown.
+     *
+     * @param times number of retries
+     * @param retryInterval pause between retries (in seconds)
+     * @param handler exception handler (receives exception on each failure)
+     * @param body operation to wrap around
+     * @return result of the body execution
+     * @throws MultipleFailureException if closure fails to execute in given amount of tries.
+     */
+    def static retry(int times = 5, double retryInterval = 2, Closure handler = { e -> log.debug(e) }, Closure body) {
+        int retries = 0
+
+        List<Throwable> ex = []
+        while (retries++ < times) {
+            try {
+                return body.call()
+            } catch (Throwable t) {
+                ex.add(t)
+                handler.call(t)
+                sleep(retryInterval * 1000 as long)
+            }
+        }
+        throw new MultipleFailureException(ex)
+    }
+
+    /**
+     * Wait for closure to return a result. Returns false if success condition was not met.
+     *
+     * @param timeout time to wait in seconds
+     * @param retryInterval pause between retries (in seconds)
+     * @param closure : A closure that returns following:
+     *          - True if check is successful
+     *          - False or null if check is unsuccessful
+     * @return True if wait was successful, false otherwise
+     */
+    static boolean wait(double timeout, double retryInterval = 0.5, Closure closure) {
+        long endTime = System.currentTimeMillis() + (long) (timeout * 1000)
+        long sleepTime = (long) (retryInterval * 1000)
+        while (System.currentTimeMillis() < endTime) {
+            try {
+                if (closure.call()) {
+                    return true
+                } else {
+                    log.debug("No wait results yet. Sleeping for ${sleepTime} ms...")
+                    sleep(sleepTime)
+                }
+            } catch (Throwable t) {
+                log.debug("Wait check threw exception: ${t.message}.\nSleeping for ${sleepTime} ms...")
+                sleep(sleepTime)
+            }
+        }
+        log.debug("Closure failed to succeed within ${timeout} seconds, aborting.")
+        return false
+    }
```

Comment:
It'd be nice to include another wait() method which would fail early if we've met a failure condition, like this:

`/**
--
* Wait for closure to return a result. Returns false if success condition was not met due to error or timeout.
*
* @param timeout time to wait in seconds
* @param retryInterval pause between retries (in seconds)
* @param closure: A tri-state closure that returns following:
*          - Object or true if check is successful
*          - False if action is failed and it doesn't make sense to wait anymore
*          - Null if success/failure criteria is not yet met
* @return Closure call result if wait was successful, false otherwise
*/
def static waitTristate(double timeout, double retryInterval = 0.5, Closure closure) {
long endTime = System.currentTimeMillis() + (long)(timeout * 1000)
long sleepTime = (long)(retryInterval * 1000)
while(System.currentTimeMillis() < endTime) {
try {
def result = closure.call()
if(result) {
return result
} else if (result == false) { //Checking if closure returned false or null.
log.debug("An error occured during wait, aborting.")
return false
} else {
log.debug("No wait results yet. Sleeping for ${sleepTime} ms...")
sleep(sleepTime)
}
} catch(Throwable t) {
log.debug("Wait check threw exception: ${t.message}.\nSleeping for ${sleepTime} ms...")
sleep(sleepTime)
}
}
log.debug("Closure failed to succeed within ${timeout} seconds, aborting.")
return false
}`

Reply:
already discussed, not going to do here